src/system_database.ts

```diff
@@ -243,17 +243,40 @@ export class PostgresSystemDatabase implements SystemDatabase {
         initStatus.recovery,
       ]
     );
-    const recovery_attempts = result.rows[0].recovery_attempts;
+    // Check the started workflow matches the expected name, class_name, config_name, and queue_name
+    // A mismatch indicates a workflow starting with the same UUID but different functions, which should not be allowed.
+    const resRow = result.rows[0];
+    resRow.config_name = resRow.config_name || "";
+    resRow.queue_name = resRow.queue_name || undefined;
+    let msg = "";
+    if (resRow.name !== initStatus.name) {
+      msg = `Workflow already exists with a different function name: ${resRow.name}, but the provided function name is: ${initStatus.name}`;
+    } else if (resRow.class_name !== initStatus.className) {
+      msg = `Workflow already exists with a different class name: ${resRow.class_name}, but the provided class name is: ${initStatus.className}`;
+    } else if (resRow.config_name !== initStatus.configName) {
+      msg = `Workflow already exists with a different config name: ${resRow.config_name}, but the provided config name is: ${initStatus.configName}`;
+    } else if (resRow.queue_name !== initStatus.queueName) {
+      msg = `Workflow already exists with a different queue name: ${resRow.queue_name}, but the provided queue name is: ${initStatus.queueName}`;
+    }
+    if (msg !== "") {
+      throw new DBOSConflictingWorkflowError(initStatus.workflowUUID, msg);
+    }
+
+    const recovery_attempts = resRow.recovery_attempts;
     if (recovery_attempts >= initStatus.maxRetries && initStatus.recovery) {
       await this.pool.query(`UPDATE ${DBOSExecutor.systemDBSchemaName}.workflow_status SET status=$1 WHERE workflow_uuid=$2 AND status=$3`, [StatusString.RETRIES_EXCEEDED, initStatus.workflowUUID, StatusString.PENDING]);
       throw new DBOSDeadLetterQueueError(initStatus.workflowUUID, initStatus.maxRetries);
     }
-    const status = result.rows[0].status;
+    const status = resRow.status;
 
+    const serializedInputs = DBOSJSON.stringify(args);
     const { rows } = await this.pool.query<workflow_inputs>(
       `INSERT INTO ${DBOSExecutor.systemDBSchemaName}.workflow_inputs (workflow_uuid, inputs) VALUES($1, $2) ON CONFLICT (workflow_uuid) DO UPDATE SET workflow_uuid = excluded.workflow_uuid  RETURNING inputs`,
-      [initStatus.workflowUUID, DBOSJSON.stringify(args)]
+      [initStatus.workflowUUID, serializedInputs]
     );
+    if (serializedInputs !== rows[0].inputs) {
+      this.logger.warn(`Workflow inputs for ${initStatus.workflowUUID} changed since the first call! Use the original inputs.`);
```

Comment:
Might be useful to print the difference. But maybe there's some privacy concerns?

Reply:
Yeah I'm worried about logging the inputs. Also the inputs could be large.